### PR TITLE
[resubmit] add replication lag and recovery time metrics

### PR DESCRIPTION
**Pull request by mstetsyuk**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add replication lag and recovery time metrics.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 69562e9c31ac4ce94c269abea7c1d9c1f65052ce with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67913/69562e9c31ac4ce94c269abea7c1d9c1f65052ce/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67913/69562e9c31ac4ce94c269abea7c1d9c1f65052ce/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Databases/DatabaseReplicated.cpp [hunk 1]
```diff
@@ -349,31 +353,50 @@ std::vector<UInt8> DatabaseReplicated::tryGetAreReplicasActive(const ClusterPtr
...
+        {
+            for (const auto & replica : addresses_with_failover[shard_index])
+            {
+                auto replica_active = zk_res[2 * global_replica_index + 1];
+                auto replica_log_ptr = zk_res[2 * global_replica_index + 2];
+
+                replicas_info[global_replica_index] = ReplicaInfo{
+                    .is_active = replica_active.error == Coordination::Error::ZOK,
+                    .replication_lag = replica_log_ptr.error != Coordination::Error::ZNONODE ? std::optional(max_log_ptr - parse<UInt32>(replica_log_ptr.data)) : std::nullopt,
+                    .recovery_time = replica.is_local && ddl_worker ? ddl_worker->getCurrentInitializationDurationMs() : 0,
```
Comment (tavplubix) [inline src/Databases/DatabaseReplicated.cpp:385]: It probably requires `ddl_worker_mutex`

File: src/Databases/DatabaseReplicated.cpp [hunk 2]
```diff
@@ -349,31 +353,50 @@ std::vector<UInt8> DatabaseReplicated::tryGetAreReplicasActive(const ClusterPtr
...
+                };
+
+                ++global_replica_index;
+            }
+        }
+
+        return replicas_info;
+    } catch (...)
```
Comment (tavplubix) [inline src/Databases/DatabaseReplicated.cpp:393]:
```suggestion
    }
    catch (...)
```

It should have been caught by Style Check; it's interesting why Style Check missed this

File: src/Storages/System/StorageSystemClusters.cpp [hunk 3]
```diff
@@ -67,6 +70,10 @@ void StorageSystemClusters::writeCluster(MutableColumns & res_columns, const std
     const auto & shards_info = cluster->getShardsInfo();
     const auto & addresses_with_failover = cluster->getShardsAddresses();
 
+    ReplicasInfo replicas_info;
+    if (replicated)
+        replicas_info = replicated->tryGetReplicasInfo(name_and_cluster.second);
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemClusters.cpp:75]: We should avoid calling it when we don't need these columns

**Review by tavplubix — APPROVED** [review]

**Pull request merged**